In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
input_file_name = "/Users/abuzatu/Work/data/finance/stocks/BullionVault/BullionVault_PL_NARIAD_10_September_2021_095253.csv"
input_file_name

'/Users/abuzatu/Work/data/finance/stocks/BullionVault/BullionVault_PL_NARIAD_10_September_2021_095253.csv'

In [3]:
df = pd.read_csv(input_file_name)
df

Asset Class           Name            Date (UTC)   Action      Order Id  \
0          GOLD    London gold  04-Dec-2011 23:28:14      BUY  1.348633e+08   
1          GOLD    London gold  04-Dec-2011 23:53:44      BUY  1.348642e+08   
2          GOLD    London gold  05-Dec-2011 11:52:54      BUY  1.348890e+08   
3          GOLD    London gold  05-Dec-2011 12:57:04      BUY  1.348917e+08   
4          GOLD    London gold  14-Dec-2011 18:34:52     SELL  1.356922e+08   
..          ...            ...                   ...      ...           ...   
794      SILVER         SILVER  01-Feb-2021 00:00:00  CUSTODY           NaN   
795      SILVER         SILVER  01-Mar-2021 00:00:00  CUSTODY           NaN   
796      SILVER         SILVER  01-Apr-2021 00:00:00  CUSTODY           NaN   
797      SILVER  London silver  26-Apr-2021 13:17:10     SELL  1.305674e+09   
798      SILVER         SILVER  01-May-2021 00:00:00  CUSTODY           NaN   

    Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
0        GBP        0.025   1058.36        NaN           NaN   
1        GBP        0.011    465.15        NaN           NaN   
2        GBP        0.016    671.70        NaN           NaN   
3        GBP        0.001     41.88        NaN           NaN   
4        GBP          NaN       NaN      0.043       1663.15   
..       ...          ...       ...        ...           ...   
794      USD          NaN      6.60        NaN           NaN   
795      USD          NaN      6.62        NaN           NaN   
796      USD          NaN      6.83        NaN           NaN   
797      EUR          NaN       NaN      3.144       2177.28   
798      USD          NaN      6.67        NaN           NaN   

     Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
0                          0.00                NaN     0.025       1041.75   
1                          0.00                NaN     0.036       1512.72   
2                          0.00                NaN     0.052       2185.04   
3                          0.00                NaN     0.053       2227.06   
4                       1815.00            -151.85     0.010        396.90   
..                          ...                ...       ...           ...   
794                        0.00                NaN     3.144       2390.38   
795                        0.00                NaN     3.144       2241.04   
796                        0.00                NaN     3.144       2107.42   
797                     1799.42             377.86     0.000          0.00   
798                        0.00                NaN     0.000          0.00   

     Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
0                 -16.61                       1058.36             42334.40   
1                 -10.79                       1523.51             42319.73   
2                 -10.17                       2195.21             42215.58   
3                 -10.03                       2237.09             42209.25   
4                 -25.19                        422.09             42209.25   
..                   ...                           ...                  ...   
794               604.41                       1785.97               568.06   
795               448.45                       1792.59               570.17   
796               308.00                       1799.42               572.34   
797                 0.00                          0.00                  NaN   
798                -6.67                          6.67                  NaN   

     Conversion Rate  
0               1.17  
1               1.16  
2               1.16  
3               1.16  
4               1.19  
..               ...  
794             0.82  
795             0.83  
796             0.85  
797             1.00  
798             0.83  

[799 rows x 18 columns]

# What the table received from BullionVault contains

Every row is the table is either:
    * a trade (buy or sell)
    * or a montly fee paid to BullionVault to hold the physical bullion (gold or silver) that I bought digitally

For every row, there is a:
* COST in euro (it includes the fee paid for each trade plus one fee every month for holding the physical bullion)
* PROCEEDS in euro

* The realised gain

# Keeping only 2020


There are 799 rows, over time.

I declared taxes in Germany in 2019, but I did not trade at BullionVault in 2019.

I now calculate for the tax declaration in Germany of calendar year 2020.

So let's keep only the subset of actions in 2020. There are 769 of them.

In [4]:
df = df[df["Date (UTC)"].str.contains("2020")]
df

Asset Class         Name            Date (UTC)   Action      Order Id  \
9          GOLD  London gold  14-Apr-2020 08:37:59      BUY  1.160904e+09   
10         GOLD  London gold  14-Apr-2020 08:37:59      BUY  1.160904e+09   
11         GOLD  London gold  14-Apr-2020 08:37:59      BUY  1.160904e+09   
12         GOLD  London gold  14-Apr-2020 09:47:54     SELL  1.160942e+09   
13         GOLD  London gold  14-Apr-2020 13:55:33     SELL  1.161043e+09   
..          ...          ...                   ...      ...           ...   
788      SILVER       SILVER  01-Aug-2020 00:00:00  CUSTODY           NaN   
789      SILVER       SILVER  01-Sep-2020 00:00:00  CUSTODY           NaN   
790      SILVER       SILVER  01-Oct-2020 00:00:00  CUSTODY           NaN   
791      SILVER       SILVER  01-Nov-2020 00:00:00  CUSTODY           NaN   
792      SILVER       SILVER  01-Dec-2020 00:00:00  CUSTODY           NaN   

    Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
9        EUR        0.007    357.45        NaN           NaN   
10       EUR        0.022   1124.30        NaN           NaN   
11       EUR        0.001     51.08        NaN           NaN   
12       EUR          NaN       NaN      0.025       1256.18   
13       EUR          NaN       NaN      0.005        254.15   
..       ...          ...       ...        ...           ...   
788      USD          NaN      6.80        NaN           NaN   
789      USD          NaN      6.71        NaN           NaN   
790      USD          NaN      6.83        NaN           NaN   
791      USD          NaN      6.87        NaN           NaN   
792      USD          NaN      6.71        NaN           NaN   

     Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
9                          0.00                NaN     0.007        354.27   
10                         0.00                NaN     0.029       1467.69   
11                         0.00                NaN     0.030       1518.30   
12                      1279.86             -23.68     0.005        253.05   
13                       255.97              -1.82     0.000          0.00   
..                          ...                ...       ...           ...   
788                        0.00                NaN     3.144       2095.48   
789                        0.00                NaN     3.144       2381.89   
790                        0.00                NaN     3.144       2013.42   
791                        0.00                NaN     3.144       2051.77   
792                        0.00                NaN     3.144       1919.73   

     Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
9                  -6.18                        360.45             51492.86   
10                -17.06                       1484.75             51198.28   
11                -17.53                       1535.83             51194.34   
12                 -2.92                        255.97             51194.34   
13                  0.00                          0.00                  NaN   
..                   ...                           ...                  ...   
788               349.78                       1745.70               555.25   
789               629.48                       1752.41               557.39   
790               254.18                       1759.24               559.56   
791               285.66                       1766.11               561.74   
792               146.91                       1772.82               563.88   

     Conversion Rate  
9               1.00  
10              1.00  
11              1.00  
12              1.00  
13              1.00  
..               ...  
788             0.85  
789             0.84  
790             0.85  
791             0.86  
792             0.84  

[769 rows x 18 columns]

Change format of date so that we can short chronologically

In [5]:
df["Date (UTC)"] = pd.to_datetime(df["Date (UTC)"])
df = df.sort_values(by="Date (UTC)")
df

<ipython-input-5-25eefdf240c8>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Date (UTC)"] = pd.to_datetime(df["Date (UTC)"])


Asset Class           Name          Date (UTC)   Action      Order Id  \
34       SILVER  London silver 2020-03-23 17:02:00      BUY  1.151656e+09   
35       SILVER  London silver 2020-03-30 00:10:05     SELL  1.154605e+09   
36       SILVER  London silver 2020-03-31 07:37:16      BUY  1.155243e+09   
37       SILVER         SILVER 2020-04-01 00:00:00  CUSTODY           NaN   
38       SILVER  London silver 2020-04-03 11:32:13     SELL  1.156931e+09   
..          ...            ...                 ...      ...           ...   
788      SILVER         SILVER 2020-08-01 00:00:00  CUSTODY           NaN   
789      SILVER         SILVER 2020-09-01 00:00:00  CUSTODY           NaN   
790      SILVER         SILVER 2020-10-01 00:00:00  CUSTODY           NaN   
791      SILVER         SILVER 2020-11-01 00:00:00  CUSTODY           NaN   
792      SILVER         SILVER 2020-12-01 00:00:00  CUSTODY           NaN   

    Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
34       EUR        3.807   1492.16        NaN           NaN   
35       GBP          NaN       NaN      3.807       1582.45   
36       GBP        3.769   1579.59        NaN           NaN   
37       USD          NaN      7.27        NaN           NaN   
38       EUR          NaN       NaN      3.769       1623.82   
..       ...          ...       ...        ...           ...   
788      USD          NaN      6.80        NaN           NaN   
789      USD          NaN      6.71        NaN           NaN   
790      USD          NaN      6.83        NaN           NaN   
791      USD          NaN      6.87        NaN           NaN   
792      USD          NaN      6.71        NaN           NaN   

     Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
34                         0.00                NaN     3.807       1423.44   
35                      1498.21              84.24     0.000          0.00   
36                         0.00                NaN     3.769       1539.26   
37                         0.00                NaN     3.769       1532.10   
38                      1586.86              36.96     0.000          0.00   
..                          ...                ...       ...           ...   
788                        0.00                NaN     3.144       2095.48   
789                        0.00                NaN     3.144       2381.89   
790                        0.00                NaN     3.144       2013.42   
791                        0.00                NaN     3.144       2051.77   
792                        0.00                NaN     3.144       1919.73   

     Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
34                -74.77                       1498.21               393.55   
35                  0.00                          0.00                  NaN   
36                -40.33                       1579.59               419.11   
37                -54.76                       1586.86               421.03   
38                  0.00                          0.00                  NaN   
..                   ...                           ...                  ...   
788               349.78                       1745.70               555.25   
789               629.48                       1752.41               557.39   
790               254.18                       1759.24               559.56   
791               285.66                       1766.11               561.74   
792               146.91                       1772.82               563.88   

     Conversion Rate  
34              1.00  
35              1.12  
36              1.12  
37              0.91  
38              1.00  
..               ...  
788             0.85  
789             0.84  
790             0.85  
791             0.86  
792             0.84  

[769 rows x 18 columns]

In [6]:
df["Asset Class"].value_counts()

SILVER    759
GOLD       10
Name: Asset Class, dtype: int64

In [7]:
df["Name"].value_counts()

London silver    750
SILVER             9
London gold        9
GOLD               1
Name: Name, dtype: int64

In [8]:
df["Action"].value_counts()

SELL       396
BUY        363
CUSTODY     10
Name: Action, dtype: int64

# Monthtly fee appear as CUSTODY

Monthly fee in EUR from BullionVault to hold the phyical bullion (silver or gold) on my behalf.

While I bought the bullion digitally, and never took physical possession, I was the actual owner of a small piece of physical bullion

In [9]:
df[df["Action"] == "CUSTODY"]

Asset Class    Name Date (UTC)   Action  Order Id Currency  Bought (kg)  \
37       SILVER  SILVER 2020-04-01  CUSTODY       NaN      USD          NaN   
433      SILVER  SILVER 2020-05-01  CUSTODY       NaN      USD          NaN   
18         GOLD    GOLD 2020-05-01  CUSTODY       NaN      USD          NaN   
730      SILVER  SILVER 2020-06-01  CUSTODY       NaN      USD          NaN   
779      SILVER  SILVER 2020-07-01  CUSTODY       NaN      USD          NaN   
788      SILVER  SILVER 2020-08-01  CUSTODY       NaN      USD          NaN   
789      SILVER  SILVER 2020-09-01  CUSTODY       NaN      USD          NaN   
790      SILVER  SILVER 2020-10-01  CUSTODY       NaN      USD          NaN   
791      SILVER  SILVER 2020-11-01  CUSTODY       NaN      USD          NaN   
792      SILVER  SILVER 2020-12-01  CUSTODY       NaN      USD          NaN   

     Cost (€)  Sold (kg)  Proceeds (€)  Bookcost of metal Sold (€)  \
37       7.27        NaN           NaN                         0.0   
433      7.33        NaN           NaN                         0.0   
18       3.66        NaN           NaN                         0.0   
730      7.21        NaN           NaN                         0.0   
779      7.13        NaN           NaN                         0.0   
788      6.80        NaN           NaN                         0.0   
789      6.71        NaN           NaN                         0.0   
790      6.83        NaN           NaN                         0.0   
791      6.87        NaN           NaN                         0.0   
792      6.71        NaN           NaN                         0.0   

     Realised Gain (€)  Kg Total  Kg Value (€)  Unrealised Gain (€)  \
37                 NaN     3.769       1532.10               -54.76   
433                NaN    10.447       4592.50               -82.04   
18                 NaN     0.000          0.00                -3.66   
730                NaN     0.000          0.00                -7.21   
779                NaN     0.000          0.00                -7.13   
788                NaN     3.144       2095.48               349.78   
789                NaN     3.144       2381.89               629.48   
790                NaN     3.144       2013.42               254.18   
791                NaN     3.144       2051.77               285.66   
792                NaN     3.144       1919.73               146.91   

     Bookcost Carried Forward (€)  Bookcost per kg (€)  Conversion Rate  
37                        1586.86               421.03             0.91  
433                       4674.54               447.46             0.91  
18                           3.66                  NaN             0.91  
730                          7.21                  NaN             0.90  
779                          7.13                  NaN             0.89  
788                       1745.70               555.25             0.85  
789                       1752.41               557.39             0.84  
790                       1759.24               559.56             0.85  
791                       1766.11               561.74             0.86  
792                       1772.82               563.88             0.84

In [10]:
# First two buys
df[df["Action"] == "BUY"][0:2]

Asset Class           Name          Date (UTC) Action      Order Id  \
34      SILVER  London silver 2020-03-23 17:02:00    BUY  1.151656e+09   
36      SILVER  London silver 2020-03-31 07:37:16    BUY  1.155243e+09   

   Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
34      EUR        3.807   1492.16        NaN           NaN   
36      GBP        3.769   1579.59        NaN           NaN   

    Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
34                         0.0                NaN     3.807       1423.44   
36                         0.0                NaN     3.769       1539.26   

    Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
34               -74.77                       1498.21               393.55   
36               -40.33                       1579.59               419.11   

    Conversion Rate  
34             1.00  
36             1.12

In [11]:
# Last two buys
df[df["Action"] == "BUY"][-2:]

Asset Class           Name          Date (UTC) Action      Order Id  \
786      SILVER  London silver 2020-07-17 01:56:30    BUY  1.190292e+09   
787      SILVER  London silver 2020-07-23 01:00:53    BUY  1.192622e+09   

    Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
786      EUR        2.725   1472.98        NaN           NaN   
787      EUR        0.419    265.92        NaN           NaN   

     Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
786                         0.0                NaN     2.725       1475.04   
787                         0.0                NaN     3.144       1993.30   

     Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
786                 2.06                       1472.98               540.55   
787               254.40                       1738.90               553.09   

     Conversion Rate  
786              1.0  
787              1.0

In [12]:
# First two sells
df[df["Action"] == "SELL"][0:2]

Asset Class           Name          Date (UTC) Action      Order Id  \
35      SILVER  London silver 2020-03-30 00:10:05   SELL  1.154605e+09   
38      SILVER  London silver 2020-04-03 11:32:13   SELL  1.156931e+09   

   Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
35      GBP          NaN       NaN      3.807       1582.45   
38      EUR          NaN       NaN      3.769       1623.82   

    Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
35                     1498.21              84.24       0.0           0.0   
38                     1586.86              36.96       0.0           0.0   

    Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
35                  0.0                           0.0                  NaN   
38                  0.0                           0.0                  NaN   

    Conversion Rate  
35             1.12  
38             1.00

In [13]:
# Last wo sells
df[df["Action"] == "SELL"][-2:]

Asset Class           Name          Date (UTC) Action      Order Id  \
784      SILVER  London silver 2020-07-06 09:09:39   SELL  1.187801e+09   
785      SILVER  London silver 2020-07-06 13:09:28   SELL  1.186875e+09   

    Currency  Bought (kg)  Cost (€)  Sold (kg)  Proceeds (€)  \
784      EUR          NaN       NaN      2.000       1038.96   
785      EUR          NaN       NaN      1.954       1017.02   

     Bookcost of metal Sold (€)  Realised Gain (€)  Kg Total  Kg Value (€)  \
784                     1026.28              12.68     1.954       1007.68   
785                     1002.67              14.35     0.000          0.00   

     Unrealised Gain (€)  Bookcost Carried Forward (€)  Bookcost per kg (€)  \
784                  5.0                       1002.67               513.14   
785                  0.0                          0.00                  NaN   

     Conversion Rate  
784              1.0  
785              1.0

# Calculate realised profit / loss

BullionVault calculates the realised gain for each row, so make the sum of all the valumes in the column "Realised Gain (€)"

In [14]:
df["Realised Gain (€)"].sum()

-248.53000000000003

# Wire charges

To this I add 20 euros in wire charges for two money withdrawals I have done in 2020. 

Please see the last page of the file `AdrianBuzatu_BullionVault_Statement_2020.pdf`.

# Conclusion

In conclusion, I would like to claim 268.53 euro as a loss from trading physical bullion (silver and gold) at BullionVault.com in calendar year 2020.